site/docs/types/named_structs.md

```diff
@@ -0,0 +1,88 @@
+# Named Structs
+
+A Named Struct is a special type construct that combines:
+* A Struct type
+* A list of names for the fields in the Struct, in depth-first search order
+
+The depth-first search order for names arises from the the ability to nest Structs within other types. All struct fields must be named, even nested fields.
+
+Named Structs are most commonly used to model the schema of Read relations.
+
+## Determining Names
+When producing/consuming names for a NamedStruct, some types requires special handling:
+
+### Struct
+A struct has names for each of its inner fields.
+
+For example, the following Struct
+```
+        a     b
+struct<i64, fp64>
+
```

Comment:
Is this the common syntax here to put the names above the types?  Or are `a` and `b` supposed to be metasyntactic variables that aren't referenced here?

Reply:
or, looking at the later examples, is the user responsible for flattening the tree of names?

Reply:
+1 for the `struct<a: i64, b: fp64>` form, maybe with bolded names

Reply:
The [NamedStruct](https://github.com/substrait-io/substrait/blob/8ae1084761c0c384c9c54bd4d7af62c4a58ea1cc/proto/substrait/type.proto#L258-L262) messages consists of a [Type.Struct](https://github.com/substrait-io/substrait/blob/8ae1084761c0c384c9c54bd4d7af62c4a58ea1cc/proto/substrait/type.proto#L195-L199) message, which only contains type information, no names, and a `names` field which is the depth-first encoded list of names associated with the struct.

> Is it this project's style to put the names above the types?

The Struct message itself does not contain any name information. The format I'm using the type names as defined in https://substrait.io/types/type_classes/#compound-types to represent the types.

I'm wary of `struct<a: i64, b: fp64>` because it makes it look like the Structs may have names. There actually is a `NSTRUCT` pseudo-construct for this, but it's not really used anywhere (that I know of).

Reply:
And changed my mind after more feedback. Added a note to the top of the page and inlined the names.

Reply:
To be clear, I wasn't arguing for a change in the internal representation... It's just that I found this particular part of Substrait quite confusing, so other passersby are likely to as well. :)

Reply:
I'm certain that it would be duplicative, but it would probably help to have a paragraph or two--or at least a _very_ specific hyperlink--about Substrait's approach here. With that context, the separate listings of types and names makes more sense...

I might even recommend going so far as to add vertical tree visualizations of the nested structures, to make the depth-first part even clearer... but pretty soon you're talking about real effort ;)